mlx/data/stream/OrderedPrefetch.cpp

```diff
@@ -0,0 +1,65 @@
+// Copyright © 2023 Apple Inc.
+
+#include "mlx/data/stream/OrderedPrefetch.h"
+
+namespace mlx {
+namespace data {
+namespace stream {
+
+OrderedPrefetch::OrderedPrefetch(
+    const std::shared_ptr<buffer::Buffer>& buffer,
+    int prefetch_size,
+    int num_thread)
+    : buffer_(buffer),
+      pool_(std::make_shared<core::ThreadPool>(num_thread)),
+      prefetchSize_(prefetch_size),
+      currentIdx_(0) {
+  if (prefetchSize_ <= 0) {
+    throw std::runtime_error(
+        "Prefetch: prefetch size must be strictly positive");
+  }
+}
+
+OrderedPrefetch::~OrderedPrefetch() {
+  std::lock_guard<std::mutex> lock(mutex_);
+
+  prefetchCache_.clear();
+}
+
+Sample OrderedPrefetch::next() const {
+  std::lock_guard<std::mutex> lock(mutex_);
+
+  // First time we are called so enqueue all the fetching
+  if (prefetchCache_.size() < prefetchSize_) {
+    for (int i = 0; i < prefetchSize_; i++) {
+      prefetchCache_.emplace_back(
+          pool_->enqueue([b = buffer_, i] { return b->get(i); }));
+    }
+  }
+
+  int64_t idx = -1;
+  if (currentIdx_ < buffer_->size()) {
+    idx = currentIdx_++;
+  }
+
+  if (idx < 0) {
+    return Sample();
+  } else {
+    int f_idx = idx % prefetchSize_;
+    std::future<Sample> fsample(std::move(prefetchCache_[f_idx]));
+    prefetchCache_[f_idx] =
+        pool_->enqueue([b = buffer_, idx] { return b->get(idx); });
+    return fsample.get();
+  }
```

Comment:
All the above is really great. I would probably unlock before calling `fsample.get()`. It probably won't make a difference in performance but there is no real reason to keep everything locked while waiting for `fsample.get()`.